File: src/org/masamotod/idea/Drupal8NamespaceDetector/Drupal8ModuleFileVisitor.java
package org.masamotod.idea.Drupal8NamespaceDetector;

import java.io.IOException;
import java.nio.file.FileVisitResult;
import java.nio.file.FileVisitor;
import java.nio.file.Path;
import java.nio.file.attribute.BasicFileAttributes;

class Drupal8ModuleFileVisitor implements FileVisitor<Path> {

  @FunctionalInterface
  interface ModuleConsumer {
    void accept(String drupalModuleName, Path drupalModulePath);
  }

  private static final String INFO_YAML_SUFFIX = ".info.yml";

  private final ModuleConsumer myAction;

  Drupal8ModuleFileVisitor(ModuleConsumer action) {
    this.myAction = action;
  }

  @Override
  public FileVisitResult preVisitDirectory(Path dir, BasicFileAttributes attrs) throws IOException {
    if (dir.getFileName().toString().startsWith(".")) {
      return FileVisitResult.SKIP_SUBTREE;
    }
    return FileVisitResult.CONTINUE;
  }

  @Override
  public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IOException {
    String filename = file.getFileName().toString();
    if (filename.endsWith(INFO_YAML_SUFFIX)) {
      String moduleName = filename.substring(0, filename.length() - INFO_YAML_SUFFIX.length());
      Path moduleDir = file.getParent();

      myAction.accept(moduleName, moduleDir);
    }

    return FileVisitResult.CONTINUE;
  }

  @Override
  public FileVisitResult visitFileFailed(Path file, IOException exc) throws IOException {
    return FileVisitResult.CONTINUE;
  }

  @Override
  public FileVisitResult postVisitDirectory(Path dir, IOException exc) throws IOException {
    return FileVisitResult.CONTINUE;
  }
}

File: src/main/kotlin/org/masamotod/idea/Drupal8NamespaceDetector/Drupal8ModuleFileVisitor.kt
package org.masamotod.idea.Drupal8NamespaceDetector

import java.io.IOException
import java.nio.file.FileVisitResult
import java.nio.file.FileVisitor
import java.nio.file.Path
import java.nio.file.attribute.BasicFileAttributes

internal class Drupal8ModuleFileVisitor(private val myAction: ModuleConsumer) : FileVisitor<Path> {
    internal fun interface ModuleConsumer {
        fun accept(drupalModuleName: String?, drupalModulePath: Path?)
    }

    @Throws(IOException::class)
    override fun preVisitDirectory(dir: Path, attrs: BasicFileAttributes): FileVisitResult {
        if (dir.fileName.toString().startsWith(".")) {
            return FileVisitResult.SKIP_SUBTREE
        }
        return FileVisitResult.CONTINUE
    }

    @Throws(IOException::class)
    override fun visitFile(file: Path, attrs: BasicFileAttributes): FileVisitResult {
        val filename = file.fileName.toString()
        if (filename.endsWith(INFO_YAML_SUFFIX)) {
            val moduleName = filename.substring(0, filename.length - INFO_YAML_SUFFIX.length)
            val moduleDir = file.parent

            myAction.accept(moduleName, moduleDir)
        }

        return FileVisitResult.CONTINUE
    }

    @Throws(IOException::class)
    override fun visitFileFailed(file: Path, exc: IOException): FileVisitResult {
        return FileVisitResult.CONTINUE
    }

    @Throws(IOException::class)
    override fun postVisitDirectory(dir: Path, exc: IOException): FileVisitResult {
        return FileVisitResult.CONTINUE
    }

    companion object {
        private const val INFO_YAML_SUFFIX = ".info.yml"
    }
}
